dashboard/app/assets/javascripts/header.js

```diff
@@ -484,3 +460,37 @@ function populateProgress(scriptName) {
     $('#level-' + appOptions.serverLevelId).parent().addClass('puzzle_outer_current');
   }
 }
+
+function renderStageProgress(stageData, progressData, clientProgress, currentLevelId) {
+  var serverProgress = progressData.levels || {};
+  var currentLevelIndex = null;
+
+  var combinedProgress = stageData.levels.map(function(level, index) {
+    if (level.id === currentLevelId) {
+      currentLevelIndex = index;
+    }
+
+    var status;
+    if (dashboard.clientState.queryParams('user_id')) {
+      // Show server progress only (the student's progress)
+      status = activityCssClass((serverProgress[level.id] || {}).result);
+    } else {
+      // Merge server progress with local progress
+      status = mergedActivityCssClass((serverProgress[level.id] || {}).result, clientProgress[level.id]);
```

Comment:
Note: In my PR, I moved some of this stuff into it's own progress namespace, so we now have dashboard.progress.activityCssClass and dashboard.progress.mergedActivityCssClass. This might result in some merge conflicts.